packages/ui-kit/src/components/AccessTokenProvider/AccessTokenProvider.tsx

```diff
@@ -0,0 +1,131 @@
+  /**
+   * If true, the access token failed to be fetched after the maximum retries
+   */
+  failedRetry?: boolean
+}
+
+const ACCESS_TOKEN_REFRESH_INTERVAL = 3300000 // 55 minutes
+const ACCESS_TOKEN_MAX_RETRIES = 3
+
+export const AccessTokenContext = createContext<AccessTokenContextValue | undefined>({})
+
+export interface AccessTokenProviderProps extends PropsWithChildren {
+  /**
+  * Function that the provider will use to fetch the access token.
+  * @returns {Promise<string>} A promise that resolves to the access token
+  */
+  fetchToken?: () => Promise<string>
+  /**
+  * If passed, the provider will ignore the `fetchToken` function and pass this access token to all the children components.
+  */
+  accessToken?: string
+  /**
+  * Function that will be called when the access token expires.
+  */
+  onAccessTokenExpired?: () => void
+}
+
+export const AccessTokenProvider = (props: AccessTokenProviderProps) => {
+  const { children, accessToken: accessTokenFromProps, fetchToken, onAccessTokenExpired } = props
+
+  const [isLoading, setIsLoading] = useState(accessTokenFromProps == null)
+  const [fetchedToken, setFetchedToken] = useState<string | undefined>(undefined)
+  const [failedRetry, setFailedRetry] = useState(false)
+
+  const log = useLog()
+
+  const shouldFetchToken = accessTokenFromProps == null
+
+  const interval = useRef<NodeJS.Timeout>()
+
+  const fetch = useCallback(async () => {
+    try {
+      const token = await fetchToken()
+      log.debug('Access token fetched successfully')
+
+      setFetchedToken(token)
+
+      return token
+    } catch (error) {
+      log.error('Failed to fetch access token', error)
+    } finally {
+      setIsLoading(false)
+    }
+  // This useCallback cannot be tiggered by `fetchToken` because it is a function
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  useEffect(() => {
+    if (shouldFetchToken) {
+      log.debug('Fetching access token')
+
+      interval.current != null && clearInterval(interval.current)
+
+      fetchedToken == null && fetch()
+
+      interval.current = setInterval(() => {
+        log.debug('Re-fetching access token after interval')
+        fetch()
+      }, ACCESS_TOKEN_REFRESH_INTERVAL)
+    }
+  // This useEffect cannot be tiggered by `log` because it is a function
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [fetch, shouldFetchToken])
+
+  const expiredTokenThrottleTimeout = useRef<NodeJS.Timeout>()
+
+  const onExpiredToken = useCallback(async () => {
+    if (!expiredTokenThrottleTimeout.current) {
+      expiredTokenThrottleTimeout.current = setTimeout(() => {
+        expiredTokenThrottleTimeout.current = undefined
+      }, 1000)
+    } else {
+      return
+    }
+
+    if (onAccessTokenExpired) {
+      onAccessTokenExpired()
+      return
+    }
+
+    log.debug('Re-fetching access token')
+
+    let retryCount = 0
+
+    while (retryCount < ACCESS_TOKEN_MAX_RETRIES) {
```

Comment:
IDK, it might make sense to retry indefinitely. I mean, otherwise, you could get into an urecoverable spot, right? After you run out of retries, what happens? Why not just continue retrying?

Reply:
Hey @markandrus , I don't think indefinitely retry is a good approach here. 

We should block it after a coupe of retry, it may led to infinity look.

Maybe we can increase it max 5, 6 or max 10

Reply:
Let's think about this in terms of the user experience we want to have. For example, I believe that, if the user's internet fails or the app's backend goes down, we should keep trying to re-fetch the access token so that, as soon as the user's internet or the app's backend is fixed, we can get a new token.

How would we achieve that user experience if we limit the number of retries?

> it would not be good to show a loader indefinitely e.g because the server is down

I agree; however, that doesn't mean the AccessTokenProvider can't retry fetching an access token, right? We can design the behavior of the Counter, etc., components to not show loaders indefinitely independently of the AccessTokenProvider's behavior.

Our main goal with AccessTokenProvider v1 should be to reduce the options and API surface area for the customer. Make it as simple as possible. Let's have a Zoom on this together tomorrow, and see what we can figure out.